// @paperback/types/lib/impl/SettingsUI/FormItemElement.js
import { Form } from './Form.js';
export function LabelRow(id, props) {
    return { ...props, id, type: 'labelRow', isHidden: props.isHidden ?? false };
}
export function InputRow(id, props) {
    return { ...props, id, type: 'inputRow', isHidden: props.isHidden ?? false };
}
export function StepperRow(id, props) {
    return {
        ...props,
        id,
        type: 'stepperRow',
        isHidden: props.isHidden ?? false,
    };
}
export function ToggleRow(id, props) {
    return { ...props, id, type: 'toggleRow', isHidden: props.isHidden ?? false };
}
export function SelectRow(id, props) {
    return { ...props, id, type: 'selectRow', isHidden: props.isHidden ?? false };
}
export function ButtonRow(id, props) {
    return { ...props, id, type: 'buttonRow', isHidden: props.isHidden ?? false };
}
export function WebViewRow(id, props) {
    return {
        ...props,
        id,
        type: 'webViewRow',
        isHidden: props.isHidden ?? false,
    };
}
export function NavigationRow(id, props) {
    return {
        ...props,
        id,
        type: 'navigationRow',
        isHidden: props.isHidden ?? false,
    };
}
export function OAuthButtonRow(id, props) {
    return {
        ...props,
        id,
        type: 'oauthButtonRow',
        isHidden: props.isHidden ?? false,
    };
}
export function DeferredItem(work) {
    return work();
}
