// @paperback/toolchain/lib/devtools/.generated/typescript/PDTLogger.js
// @generated by protobuf-ts 2.9.6 with parameter client_grpc1,ts_nocheck
// @generated from protobuf file "PDTLogger.proto" (syntax proto3)
// tslint:disable
// @ts-nocheck
import { ServiceType } from '@protobuf-ts/runtime-rpc';
import { WireType } from '@protobuf-ts/runtime';
import { UnknownFieldHandler } from '@protobuf-ts/runtime';
import { reflectionMergePartial } from '@protobuf-ts/runtime';
import { MessageType } from '@protobuf-ts/runtime';
import { Timestamp } from './google/protobuf/timestamp.js';
/**
 * @generated from protobuf enum LogLevel
 */
export var LogLevel;
(function (LogLevel) {
    /**
     * @generated from protobuf enum value: INFO = 0;
     */
    LogLevel[LogLevel["INFO"] = 0] = "INFO";
    /**
     * @generated from protobuf enum value: WARN = 1;
     */
    LogLevel[LogLevel["WARN"] = 1] = "WARN";
    /**
     * @generated from protobuf enum value: ERROR = 2;
     */
    LogLevel[LogLevel["ERROR"] = 2] = "ERROR";
})(LogLevel || (LogLevel = {}));
// @generated message type with reflection information, may provide speed optimized methods
class LogLine$Type extends MessageType {
    constructor() {
        super('LogLine', [
            { no: 1, name: 'level', kind: 'enum', T: () => ['LogLevel', LogLevel] },
            {
                no: 2,
                name: 'tags',
                kind: 'scalar',
                repeat: 2 /*RepeatType.UNPACKED*/,
                T: 9 /*ScalarType.STRING*/,
            },
            { no: 3, name: 'message', kind: 'scalar', T: 9 /*ScalarType.STRING*/ },
            { no: 4, name: 'date', kind: 'message', T: () => Timestamp },
        ]);
    }
    create(value) {
        const message = globalThis.Object.create(this.messagePrototype);
        message.level = 0;
        message.tags = [];
        message.message = '';
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        const message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            const [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* LogLevel level */ 1:
                    message.level = reader.int32();
                    break;
                case /* repeated string tags */ 2:
                    message.tags.push(reader.string());
                    break;
                case /* string message */ 3:
                    message.message = reader.string();
                    break;
                case /* google.protobuf.Timestamp date */ 4:
                    message.date = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.date);
                    break;
                default:
                    const u = options.readUnknownField;
                    if (u === 'throw')
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    const d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* LogLevel level = 1; */
        if (message.level !== 0)
            writer.tag(1, WireType.Varint).int32(message.level);
        /* repeated string tags = 2; */
        for (let i = 0; i < message.tags.length; i++)
            writer.tag(2, WireType.LengthDelimited).string(message.tags[i]);
        /* string message = 3; */
        if (message.message !== '')
            writer.tag(3, WireType.LengthDelimited).string(message.message);
        /* google.protobuf.Timestamp date = 4; */
        if (message.date)
            Timestamp.internalBinaryWrite(message.date, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
        const u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message LogLine
 */
export const LogLine = new LogLine$Type();
// @generated message type with reflection information, may provide speed optimized methods
class LogFilter$Type extends MessageType {
    constructor() {
        super('LogFilter', []);
    }
    create(value) {
        const message = globalThis.Object.create(this.messagePrototype);
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        const message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            const [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                default:
                    const u = options.readUnknownField;
                    if (u === 'throw')
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    const d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        const u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message LogFilter
 */
export const LogFilter = new LogFilter$Type();
/**
 * @generated ServiceType for protobuf service PaperbackLogger
 */
export const PaperbackLogger = new ServiceType('PaperbackLogger', [
    {
        name: 'streamLogs',
        serverStreaming: true,
        options: {},
        I: LogFilter,
        O: LogLine,
    },
]);
