// @paperback/toolchain/lib/devtools/.generated/typescript/PDTSourceTester.js
// @generated by protobuf-ts 2.9.6 with parameter client_grpc1,ts_nocheck
// @generated from protobuf file "PDTSourceTester.proto" (syntax proto3)
// tslint:disable
// @ts-nocheck
import { ServiceType } from '@protobuf-ts/runtime-rpc';
import { WireType } from '@protobuf-ts/runtime';
import { UnknownFieldHandler } from '@protobuf-ts/runtime';
import { reflectionMergePartial } from '@protobuf-ts/runtime';
import { MessageType } from '@protobuf-ts/runtime';
import { Timestamp } from './google/protobuf/timestamp.js';
// @generated message type with reflection information, may provide speed optimized methods
class SourceInstallRequest$Type extends MessageType {
    constructor() {
        super('SourceInstallRequest', [
            { no: 1, name: 'sourceId', kind: 'scalar', T: 9 /*ScalarType.STRING*/ },
            {
                no: 2,
                name: 'repoBaseUrl',
                kind: 'scalar',
                T: 9 /*ScalarType.STRING*/,
            },
        ]);
    }
    create(value) {
        const message = globalThis.Object.create(this.messagePrototype);
        message.sourceId = '';
        message.repoBaseUrl = '';
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        const message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            const [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* string sourceId */ 1:
                    message.sourceId = reader.string();
                    break;
                case /* string repoBaseUrl */ 2:
                    message.repoBaseUrl = reader.string();
                    break;
                default:
                    const u = options.readUnknownField;
                    if (u === 'throw')
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    const d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* string sourceId = 1; */
        if (message.sourceId !== '')
            writer.tag(1, WireType.LengthDelimited).string(message.sourceId);
        /* string repoBaseUrl = 2; */
        if (message.repoBaseUrl !== '')
            writer.tag(2, WireType.LengthDelimited).string(message.repoBaseUrl);
        const u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message SourceInstallRequest
 */
export const SourceInstallRequest = new SourceInstallRequest$Type();
// @generated message type with reflection information, may provide speed optimized methods
class SourceInstallResponse$Type extends MessageType {
    constructor() {
        super('SourceInstallResponse', []);
    }
    create(value) {
        const message = globalThis.Object.create(this.messagePrototype);
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        const message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            const [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                default:
                    const u = options.readUnknownField;
                    if (u === 'throw')
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    const d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        const u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message SourceInstallResponse
 */
export const SourceInstallResponse = new SourceInstallResponse$Type();
// @generated message type with reflection information, may provide speed optimized methods
class SearchData$Type extends MessageType {
    constructor() {
        super('SearchData', [
            {
                no: 1,
                name: 'query',
                kind: 'scalar',
                opt: true,
                T: 9 /*ScalarType.STRING*/,
            },
            {
                no: 2,
                name: 'includedTags',
                kind: 'scalar',
                repeat: 2 /*RepeatType.UNPACKED*/,
                T: 9 /*ScalarType.STRING*/,
            },
            {
                no: 3,
                name: 'excludedTags',
                kind: 'scalar',
                repeat: 2 /*RepeatType.UNPACKED*/,
                T: 9 /*ScalarType.STRING*/,
            },
            {
                no: 4,
                name: 'itemIndex',
                kind: 'scalar',
                opt: true,
                T: 9 /*ScalarType.STRING*/,
            },
            {
                no: 5,
                name: 'itemId',
                kind: 'scalar',
                opt: true,
                T: 9 /*ScalarType.STRING*/,
            },
        ]);
    }
    create(value) {
        const message = globalThis.Object.create(this.messagePrototype);
        message.includedTags = [];
        message.excludedTags = [];
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        const message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            const [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* optional string query */ 1:
                    message.query = reader.string();
                    break;
                case /* repeated string includedTags */ 2:
                    message.includedTags.push(reader.string());
                    break;
                case /* repeated string excludedTags */ 3:
                    message.excludedTags.push(reader.string());
                    break;
                case /* optional string itemIndex */ 4:
                    message.itemIndex = reader.string();
                    break;
                case /* optional string itemId */ 5:
                    message.itemId = reader.string();
                    break;
                default:
                    const u = options.readUnknownField;
                    if (u === 'throw')
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    const d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* optional string query = 1; */
        if (message.query !== undefined)
            writer.tag(1, WireType.LengthDelimited).string(message.query);
        /* repeated string includedTags = 2; */
        for (let i = 0; i < message.includedTags.length; i++)
            writer.tag(2, WireType.LengthDelimited).string(message.includedTags[i]);
        /* repeated string excludedTags = 3; */
        for (let i = 0; i < message.excludedTags.length; i++)
            writer.tag(3, WireType.LengthDelimited).string(message.excludedTags[i]);
        /* optional string itemIndex = 4; */
        if (message.itemIndex !== undefined)
            writer.tag(4, WireType.LengthDelimited).string(message.itemIndex);
        /* optional string itemId = 5; */
        if (message.itemId !== undefined)
            writer.tag(5, WireType.LengthDelimited).string(message.itemId);
        const u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message SearchData
 */
export const SearchData = new SearchData$Type();
// @generated message type with reflection information, may provide speed optimized methods
class TestData$Type extends MessageType {
    constructor() {
        super('TestData', [
            {
                no: 1,
                name: 'mangaId',
                kind: 'scalar',
                opt: true,
                T: 9 /*ScalarType.STRING*/,
            },
            {
                no: 2,
                name: 'chapterId',
                kind: 'scalar',
                opt: true,
                T: 9 /*ScalarType.STRING*/,
            },
            { no: 3, name: 'updateTime', kind: 'message', T: () => Timestamp },
            { no: 4, name: 'searchData', kind: 'message', T: () => SearchData },
        ]);
    }
    create(value) {
        const message = globalThis.Object.create(this.messagePrototype);
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        const message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            const [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* optional string mangaId */ 1:
                    message.mangaId = reader.string();
                    break;
                case /* optional string chapterId */ 2:
                    message.chapterId = reader.string();
                    break;
                case /* optional google.protobuf.Timestamp updateTime */ 3:
                    message.updateTime = Timestamp.internalBinaryRead(reader, reader.uint32(), options, message.updateTime);
                    break;
                case /* optional SearchData searchData */ 4:
                    message.searchData = SearchData.internalBinaryRead(reader, reader.uint32(), options, message.searchData);
                    break;
                default:
                    const u = options.readUnknownField;
                    if (u === 'throw')
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    const d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* optional string mangaId = 1; */
        if (message.mangaId !== undefined)
            writer.tag(1, WireType.LengthDelimited).string(message.mangaId);
        /* optional string chapterId = 2; */
        if (message.chapterId !== undefined)
            writer.tag(2, WireType.LengthDelimited).string(message.chapterId);
        /* optional google.protobuf.Timestamp updateTime = 3; */
        if (message.updateTime)
            Timestamp.internalBinaryWrite(message.updateTime, writer.tag(3, WireType.LengthDelimited).fork(), options).join();
        /* optional SearchData searchData = 4; */
        if (message.searchData)
            SearchData.internalBinaryWrite(message.searchData, writer.tag(4, WireType.LengthDelimited).fork(), options).join();
        const u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message TestData
 */
export const TestData = new TestData$Type();
// @generated message type with reflection information, may provide speed optimized methods
class SourceTestRequest$Type extends MessageType {
    constructor() {
        super('SourceTestRequest', [
            { no: 1, name: 'sourceId', kind: 'scalar', T: 9 /*ScalarType.STRING*/ },
            { no: 2, name: 'data', kind: 'message', T: () => TestData },
        ]);
    }
    create(value) {
        const message = globalThis.Object.create(this.messagePrototype);
        message.sourceId = '';
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        const message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            const [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* string sourceId */ 1:
                    message.sourceId = reader.string();
                    break;
                case /* TestData data */ 2:
                    message.data = TestData.internalBinaryRead(reader, reader.uint32(), options, message.data);
                    break;
                default:
                    const u = options.readUnknownField;
                    if (u === 'throw')
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    const d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* string sourceId = 1; */
        if (message.sourceId !== '')
            writer.tag(1, WireType.LengthDelimited).string(message.sourceId);
        /* TestData data = 2; */
        if (message.data)
            TestData.internalBinaryWrite(message.data, writer.tag(2, WireType.LengthDelimited).fork(), options).join();
        const u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message SourceTestRequest
 */
export const SourceTestRequest = new SourceTestRequest$Type();
// @generated message type with reflection information, may provide speed optimized methods
class SourceTestResponse$Type extends MessageType {
    constructor() {
        super('SourceTestResponse', [
            { no: 1, name: 'testCase', kind: 'scalar', T: 9 /*ScalarType.STRING*/ },
            {
                no: 2,
                name: 'completeTime',
                kind: 'scalar',
                T: 4 /*ScalarType.UINT64*/,
                L: 0 /*LongType.BIGINT*/,
            },
            {
                no: 4,
                name: 'failures',
                kind: 'scalar',
                repeat: 2 /*RepeatType.UNPACKED*/,
                T: 9 /*ScalarType.STRING*/,
            },
        ]);
    }
    create(value) {
        const message = globalThis.Object.create(this.messagePrototype);
        message.testCase = '';
        message.completeTime = 0n;
        message.failures = [];
        if (value !== undefined)
            reflectionMergePartial(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        const message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            const [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* string testCase */ 1:
                    message.testCase = reader.string();
                    break;
                case /* uint64 completeTime */ 2:
                    message.completeTime = reader.uint64().toBigInt();
                    break;
                case /* repeated string failures */ 4:
                    message.failures.push(reader.string());
                    break;
                default:
                    const u = options.readUnknownField;
                    if (u === 'throw')
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    const d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* string testCase = 1; */
        if (message.testCase !== '')
            writer.tag(1, WireType.LengthDelimited).string(message.testCase);
        /* uint64 completeTime = 2; */
        if (message.completeTime !== 0n)
            writer.tag(2, WireType.Varint).uint64(message.completeTime);
        /* repeated string failures = 4; */
        for (let i = 0; i < message.failures.length; i++)
            writer.tag(4, WireType.LengthDelimited).string(message.failures[i]);
        const u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message SourceTestResponse
 */
export const SourceTestResponse = new SourceTestResponse$Type();
/**
 * @generated ServiceType for protobuf service PaperbackSourceTester
 */
export const PaperbackSourceTester = new ServiceType('PaperbackSourceTester', [
    {
        name: 'installSource',
        options: {},
        I: SourceInstallRequest,
        O: SourceInstallResponse,
    },
    {
        name: 'testSource',
        serverStreaming: true,
        options: {},
        I: SourceTestRequest,
        O: SourceTestResponse,
    },
]);
