// @paperback/types/lib/impl/BasicRateLimiter.js
import { lock, unlock } from './Lock.js';
import { PaperbackInterceptor } from './PaperbackInterceptor.js';
export class BasicRateLimiter extends PaperbackInterceptor {
    options;
    promise;
    currentRequestsMade = 0;
    lastReset = Date.now();
    imageRegex = new RegExp(/\.(png|gif|jpeg|jpg|webp)(\?|$)/i);
    constructor(id, options) {
        super(id);
        this.options = options;
    }
    async interceptRequest(request) {
        if (this.options.ignoreImages && this.imageRegex.test(request.url)) {
            return request;
        }
        await lock(this.id);
        await this.incrementRequestCount();
        unlock(this.id);
        return request;
    }
    async interceptResponse(request, response, data) {
        return data;
    }
    async incrementRequestCount() {
        await this.promise;
        const secondsSinceLastReset = (Date.now() - this.lastReset) / 1000;
        if (secondsSinceLastReset > this.options.bufferInterval) {
            this.currentRequestsMade = 0;
            this.lastReset = Date.now();
        }
        this.currentRequestsMade += 1;
        if (this.currentRequestsMade >= this.options.numberOfRequests) {
            const secondsSinceLastReset = (Date.now() - this.lastReset) / 1000;
            if (secondsSinceLastReset <= this.options.bufferInterval) {
                const sleepTime = this.options.bufferInterval - secondsSinceLastReset;
                console.log(`[BasicRateLimiter] rate limit hit, sleeping for ${sleepTime}`);
                this.promise = Application.sleep(sleepTime);
            }
        }
    }
}
