// @paperback/types/lib/impl/SourceInfo.js
/* eslint-disable @typescript-eslint/prefer-literal-enum-member */
export var SourceIntents;
(function (SourceIntents) {
    SourceIntents[SourceIntents["NONE"] = 0] = "NONE";
    /**
     * @deprecated use {@link CHAPTER_PROVIDING}
     */
    SourceIntents[SourceIntents["MANGA_CHAPTERS"] = 1] = "MANGA_CHAPTERS";
    SourceIntents[SourceIntents["CHAPTER_PROVIDING"] = 1] = "CHAPTER_PROVIDING";
    /**
     * @deprecated use {@link MANGA_PROGRESS_PROVIDING}
     */
    SourceIntents[SourceIntents["MANGA_PROGRESS"] = 2] = "MANGA_PROGRESS";
    SourceIntents[SourceIntents["MANGA_PROGRESS_PROVIDING"] = 2] = "MANGA_PROGRESS_PROVIDING";
    /**
     * @deprecated use {@link DISCOVER_SECIONS_PROVIDING}
     */
    SourceIntents[SourceIntents["DISCOVER_SECIONS"] = 4] = "DISCOVER_SECIONS";
    SourceIntents[SourceIntents["DISCOVER_SECIONS_PROVIDING"] = 4] = "DISCOVER_SECIONS_PROVIDING";
    /**
     * @deprecated use {@link MANAGED_COLLECTION_PROVIDING}
     */
    SourceIntents[SourceIntents["COLLECTION_MANAGEMENT"] = 8] = "COLLECTION_MANAGEMENT";
    SourceIntents[SourceIntents["MANAGED_COLLECTION_PROVIDING"] = 8] = "MANAGED_COLLECTION_PROVIDING";
    /**
     * @deprecated use {@link CLOUDFLARE_BYPASS_PROVIDING}
     */
    SourceIntents[SourceIntents["CLOUDFLARE_BYPASS_REQUIRED"] = 16] = "CLOUDFLARE_BYPASS_REQUIRED";
    SourceIntents[SourceIntents["CLOUDFLARE_BYPASS_PROVIDING"] = 16] = "CLOUDFLARE_BYPASS_PROVIDING";
    /**
     * @deprecated use {@link SETTINGS_FORM_PROVIDING}
     */
    SourceIntents[SourceIntents["SETTINGS_UI"] = 32] = "SETTINGS_UI";
    SourceIntents[SourceIntents["SETTINGS_FORM_PROVIDING"] = 32] = "SETTINGS_FORM_PROVIDING";
    /**
     * @deprecated use {@link SEARCH_RESULTS_PROVIDING}
     */
    SourceIntents[SourceIntents["MANGA_SEARCH"] = 64] = "MANGA_SEARCH";
    SourceIntents[SourceIntents["SEARCH_RESULTS_PROVIDING"] = 64] = "SEARCH_RESULTS_PROVIDING";
})(SourceIntents || (SourceIntents = {}));
/**
 * A content rating to be attributed to each source.
 */
export var ContentRating;
(function (ContentRating) {
    ContentRating["EVERYONE"] = "SAFE";
    ContentRating["MATURE"] = "MATURE";
    ContentRating["ADULT"] = "ADULT";
})(ContentRating || (ContentRating = {}));
