// @paperback/runtime-polyfills/lib/SelectorRegistry.js
/* eslint-disable @typescript-eslint/no-explicit-any */
import {} from '@paperback/types';
class SelectorRef {
    key;
    objRef;
    get value() {
        const obj = this.objRef.deref();
        if (!obj) {
            return undefined;
        }
        const value = obj[this.key];
        if (typeof value === 'function') {
            return value.bind(obj);
        }
        else {
            return value;
        }
    }
    constructor(obj, key) {
        this.key = key;
        this.objRef = new WeakRef(obj);
    }
}
export class MockSelectorRegistry {
    registry = {};
    registerSelector(id, obj, key) {
        this.registry[id] = new SelectorRef(obj, key);
    }
    unregisterSelector(id) {
        // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
        delete this.registry[id];
    }
    selector(id) {
        return this.registry[id]?.value;
    }
    Selector = (obj, symbol) => {
        function relative_uuid() {
            return `${Date.now()}-${Math.random().toString().slice(2)}`;
        }
        const canonicalId = (function (obj, symbol) {
            const key = `$__selector_${String(symbol)}`;
            const existingCanonicalId = obj[key];
            if (existingCanonicalId) {
                return existingCanonicalId;
            }
            else {
                const id = relative_uuid();
                Object.defineProperty(obj, key, { enumerable: true, value: id });
                return id;
            }
        })(obj, symbol);
        this.registerSelector(canonicalId, obj, symbol);
        return canonicalId;
    };
}
